### PR TITLE
Python script to generate tabby file table from directory

### PR DESCRIPTION
**Pull request by jsheunis**
This PR:
- Is a pure Python script that prints a table with information about the files contained in a directory on a filesystem. File information includes: file path, file content checksum (using a specified hash algorithm), file size in bytes, and an empty url column. The table can be output to stdout or to a text file in TSV format. The script recurses into subdirectories by default, although this can be switched off with a flag.
- Complies with tabby convention `@tby-ds1` (https://docs.datalad.org/projects/tabby/en/latest/conventions/tby-ds1.html) i terms of the name of output TSV file as well as the header names of the columns in the TSV file / table.
- Closes https://github.com/psychoinformatics-de/datalad-tabby/issues/33

Possible further enhancements:
- additional arguments: levels of recursion, file exclusion regex.
- In order to make this executable on multiple OSs without requiring Python installation, we could consider pyinstaller and running a github action or appveyor workflow to generate executables for the different operating systems.

Example usage:

```
# print table of files in current working directory (recursively) to STDOUT
python tools/dir2filetable.py .

# print table of files in current working directory (recursively) to a TSV file in an existing directory
# this prints to mydir/files@tby-ds1.tsv
python tools/dir2filetable.py --output <mydir> .

# print table of files in current working directory (recursively) to a TSV file with a named prefix
# this prints to mydir/myprefix_files@tby-ds1.tsv
python tools/dir2filetable.py --output <mydir>/<myprefix> .

# print table of files in current working directory NON-recursively to STDOUT
python tools/dir2filetable.py --non-recursive .
```

### CODE REVIEW

**Review by mih — CHANGES_REQUESTED** [review]
Thx, this works for me!

I would suggest to make the following changes.

- Move it out of the package source tree. it is a standalone script that could be downloaded individually (which is the main benefit, because package installation pull in datalad, but see #95(. I suggest `tools/dir2filetable.py`.
- add a header row. This is needed to be compliant with the `many` table layout. If this is added, we may need to rethink the default for `--recursive`. I'd set it to `True` to get the full table -- as ooposed to suggesting to build many table and merge them. This may indicate that added a column header should be optional (and likely enabled by default).
- consider adopting/aligning with the `tby-ds1` convention (see #101), which would specify the column headers and semantics
- if `tby-ds1` is adopted, consider writing the content out into a file with the correct name `<prefix>[/_]files@tby-ds1.tsv`

**Comment by jsheunis** [comment]
Thanks for the useful feedback 👍

**Comment by jsheunis** [comment]
> add a header row. This is needed to be compliant with the many table layout. If this is added, we may need to rethink the default for --recursive. I'd set it to True to get the full table -- as ooposed to suggesting to build many table and merge them. This may indicate that added a column header should be optional (and likely enabled by default).

TBC: the header is already added to the file output, just not to the stdout (yet). The header will be changed to be compliant with `@tby-ds1` convention. But I don't see how adding a header has influence on the recursive process, since adding the header is done once at the output stage, after all possible recursion or non-recursion has been completed.

Regarding doing recursion by default, that seems sensible. The options IMO would then either be:
- change the current flag to something like `--no-recursion` with `action=store_false` (default `False`), or
- change the current flag to `--recursion-level` an integer which defaults to infinity
Any preference or alternatives?

File: tools/dir2filetable.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    nonrecursive: bool = False,
+    output: str = 'stdout',
+):
+    """
+    Print a table with information about the files contained in
+    a directory on a filesystem.
+
+    Args:
+        rootpath (str): The root directory for which the table 
+            shoult be printed.
```
Comment (jsheunis) [inline tools/dir2filetable.py:34]:
```suggestion
            should be printed.
```

File: tools/dir2filetable.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+from pathlib import Path
+import argparse
+import hashlib
+import csv
+
+
+def dir2filetable(
+    rootpath: str,
+    hash: str = 'md5',
+    nonrecursive: bool = False,
```
Comment (mih) [inline tools/dir2filetable.py:25]:
```suggestion
    recursive: bool = True,
```

File: tools/dir2filetable.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        
+        recursive (bool, optional): Set this flag to recurse into
+            subdirectories. Defaults to False.
+
+        output (str, optional): A directory of filename to save the
+            output to. Defaults to 'stdout.
+    """
+    # Get file list
+    out_info = []
+    _dir2filelist(Path(rootpath), Path(rootpath), out_info, hash, not nonrecursive)
```
Comment (mih) [inline tools/dir2filetable.py:49]:
```suggestion
    _dir2filelist(Path(rootpath), Path(rootpath), out_info, hash, recursive)
```

File: tools/dir2filetable.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        "If a filename supplied, it will be used to derive the prefix which is "
+        "then prepended to '_files@tby-ds1.tsv' to form the filename at which the "
+        "output will be saved. "
+        "If this argument is not supplied, the output is printed to STDOUT"
+    )
+    args = p.parse_args()
+    dir2filetable(
+        rootpath=args.directory,
+        hash=args.hash,
+        nonrecursive=args.non_recursive,
```
Comment (mih) [inline tools/dir2filetable.py:179]:
```suggestion
        recursive=not args.non_recursive,
```

**Review by mih — CHANGES_REQUESTED** [review]
Thanks, looks good. I made a series of suggested to keep the not/non recursion business a CLI problem.

**Review by mih — APPROVED** [review]

**Pull request merged**